pilot/pkg/proxy/envoy/v2/adapter.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	corev2 "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	corev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	ads "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+)
+
+// DiscoveryStreamV2Adapter is a DiscoveryStream that converts v3 Discovery messages to v2 messages.
+// This allows v2 clients to connect, while keeping the core logic of Pilot in v3 XDS.
+// This comes with a very small performance impact, as most resources are directly copied over - most
+// importantly the actual XDS response objects are untouched.
+// Note: this is just about the transport protocol. The XDS resource versioning is independent of transport
+// protocol
+type DiscoveryStreamV2Adapter struct {
+	ads.AggregatedDiscoveryService_StreamAggregatedResourcesServer
+}
+
+// We implement the v3 DiscoveryStream API
+var _ DiscoveryStream = &DiscoveryStreamV2Adapter{}
+
+func (d DiscoveryStreamV2Adapter) Send(r *discovery.DiscoveryResponse) error {
+	return d.AggregatedDiscoveryService_StreamAggregatedResourcesServer.Send(&xdsapi.DiscoveryResponse{
+		VersionInfo: r.VersionInfo,
+		Resources:   r.Resources,
+		Canary:      r.Canary,
+		TypeUrl:     r.TypeUrl,
+		Nonce:       r.Nonce,
+	})
+}
+
+func (d DiscoveryStreamV2Adapter) Recv() (*discovery.DiscoveryRequest, error) {
+	v2Req, err := d.AggregatedDiscoveryService_StreamAggregatedResourcesServer.Recv()
+	if v2Req == nil {
+		return nil, err
+	}
+	return &discovery.DiscoveryRequest{
+		VersionInfo:   v2Req.VersionInfo,
+		Node:          convertToV3Node(v2Req.Node),
+		ResourceNames: v2Req.ResourceNames,
+		TypeUrl:       v2Req.TypeUrl,
+		ResponseNonce: v2Req.ResponseNonce,
+		ErrorDetail:   v2Req.ErrorDetail,
+	}, err
+}
+
+func convertToV3Node(node *corev2.Node) *corev3.Node {
+	if node == nil {
+		return nil
+	}
+	resp := &corev3.Node{
+		Id:             node.Id,
+		Cluster:        node.Cluster,
+		Metadata:       node.Metadata,
+		ClientFeatures: node.ClientFeatures,
```

Comment:
do we care about ClientFeatures? If we are including for completeness, I think we should also add `extensions`, `user_agent_name` etc

Reply:
those other two require a deep copy which seemed excessive since we don't use them. Client features is a shallow copy. We don't use it anyways though so I would be find with removal